### PR TITLE
C2PA-55: Use of the `C2PA` font table

### PR DESCRIPTION
**Pull request by brogdonm**
# JIRA Issue

- https://monotype.atlassian.net/browse/C2PA-55

# Checklist

- [x] **PR labeled** appropriately
- [ ] **Documentation** updated:
  - [ ] Developer documentation (ReadMe files)
  - [ ] Product documentation (Release notes, PDK Guide, Functional Spec, API documents, ...)
- [X] **Test case(s)** added
  - [X] Unit Tests

# Notes for Reviewers

Updates to the latest version `0.19.1` from upstream and adds support for `C2PA` font table for storing the manifest store and uses the remote manifest uri. The changes support file and stream support. I'm still not sure how to get the remote manifest URI, but it is embedded as well.

Note, this also adds support for remote manifests URIs.

# Verification Instructions

Can you still sign a font?

Can you add the URI to the font?

Use the changes in https://github.com/Monotype/c2patool/pull/4 for verifying.

Since the `"otf"` feature is optional and off by default, to run the unit tests added you will need to run one of the following:

- `cargo test --lib asset_handlers::otf_io --features "otf, file_io"` - to run only the `otf_io` unit tests
- `cargo test --all-features` - to run all unit tests, including `otf_io` unit tests.

> Actively maintained by the @Monotype/driverpdldev team.

### CODE REVIEW

File: rustfmt.toml [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
+# Invoke by using:
+#
+# rustup toolchain add nightly
+# cargo +nightly fmt
+
+blank_lines_upper_bound = 1
 edition = "2018"
+format_code_in_doc_comments = true
 group_imports = "StdExternalCrate"
+hex_literal_case = "Lower"
```
Comment (chriskushmonotype) [inline rustfmt.toml:13]: 🕶️ I'm in!

File: sdk/src/assertions/schema_org.rs [hunk 2]
```diff
@@ -134,10 +134,10 @@ impl AssertionBase for SchemaDotOrg {
 pub struct SchemaDotOrgPerson(SchemaDotOrg);
 
 impl SchemaDotOrgPerson {
-    pub const PERSON: &'static str = "Person";
-    pub const NAME: &'static str = "name";
-    pub const IDENTIFIER: &'static str = "identifier";
     pub const CREDENTIAL: &'static str = "credential";
+    pub const IDENTIFIER: &'static str = "identifier";
+    pub const NAME: &'static str = "name";
+    pub const PERSON: &'static str = "Person";
```
Comment (chriskushmonotype) [inline sdk/src/assertions/schema_org.rs:140]: 🔢

File: sdk/src/asset_handlers/bmff_io.rs [hunk 3]
```diff
@@ -45,7 +45,7 @@ const HEADER_SIZE: u64 = 8; // 4 byte type + 4 byte size
 const HEADER_SIZE_LARGE: u64 = 16; // 4 byte type + 4 byte size + 8 byte large size
 
 const C2PA_UUID: [u8; 16] = [
-    0xD8, 0xFE, 0xC3, 0xD6, 0x1B, 0x0E, 0x48, 0x3C, 0x92, 0x97, 0x58, 0x28, 0x87, 0x7E, 0xC4, 0x81,
+    0xd8, 0xfe, 0xc3, 0xd6, 0x1b, 0x0e, 0x48, 0x3c, 0x92, 0x97, 0x58, 0x28, 0x87, 0x7e, 0xc4, 0x81,
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/bmff_io.rs:48]: 🕶 a black car with some doors that open the wrong way

File: sdk/src/asset_handlers/bmff_io.rs [hunk 4]
```diff
@@ -690,16 +691,16 @@ fn adjust_known_offsets<W: Write + CAIRead>(
             output.read_exact(&mut iloc_header)?;
 
             // get offset size (high nibble)
-            let offset_size: u8 = (iloc_header[0] & 0xF0) >> 4;
+            let offset_size: u8 = (iloc_header[0] & 0xf0) >> 4;
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/bmff_io.rs:694]: 🕶 the one member of the group in a **_white_** suit.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 5]
```diff
@@ -690,16 +691,16 @@ fn adjust_known_offsets<W: Write + CAIRead>(
...
 
             // get length size (low nibble)
-            let length_size: u8 = iloc_header[0] & 0x0F;
+            let length_size: u8 = iloc_header[0] & 0x0f;
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/bmff_io.rs:697]: 🕶 a pince-nez that is somehow cool, not fussy.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 6]
```diff
@@ -690,16 +691,16 @@ fn adjust_known_offsets<W: Write + CAIRead>(
...
 
             // get box offset size (high nibble)
-            let base_offset_size: u8 = (iloc_header[1] & 0xF0) >> 4;
+            let base_offset_size: u8 = (iloc_header[1] & 0xf0) >> 4;
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/bmff_io.rs:700]: 🕶 turns out that helicopters are pretty easy to fly.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 7]
```diff
@@ -690,16 +691,16 @@ fn adjust_known_offsets<W: Write + CAIRead>(
...
 
             // get index size (low nibble)
-            let index_size: u8 = iloc_header[1] & 0x0F;
+            let index_size: u8 = iloc_header[1] & 0x0f;
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/bmff_io.rs:703]: 🕶 spoons all flopping around.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 8]
```diff
@@ -731,7 +732,7 @@ fn adjust_known_offsets<W: Write + CAIRead>(
                     output.read_exact(&mut cm_bytes)?;
 
                     // lower nibble of 2nd byte
-                    cm_bytes[1] & 0x0F
+                    cm_bytes[1] & 0x0f
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/bmff_io.rs:735]: 🕶 "Not too bright, though."

File: sdk/src/claim.rs [hunk 9]
```diff
@@ -247,8 +247,8 @@ pub enum AssertionStoreJsonFormat {
 pub enum RemoteManifest {
     NoRemote,                // No external manifest (default)
     SideCar,        // Manifest will be saved as a side car file, output asset is untouched.
-    Remote(String), // Manifest will be saved as a side car file, output asset will contain remote reference
-    EmbedWithRemote(String), // Manifest will be embedded with a remote reference, sidecar will be generated
+    Remote(String), /* Manifest will be saved as a side car file, output asset will contain remote reference */
+    EmbedWithRemote(String), /* Manifest will be embedded with a remote reference, sidecar will be generated */
```
Comment (chriskushmonotype) [inline sdk/src/claim.rs:251]: Heh? Why tho?

File: sdk/src/error.rs [hunk 10]
```diff
@@ -90,9 +90,9 @@ pub enum Error {
     /// The COSE Sign1 structure can not be parsed.
     #[error("COSE Sign1 structure can not be parsed: {coset_error}")]
     InvalidCoseSignature {
-        coset_error: coset::CoseError, // NOTE: We can not use #[transparent] here because
-                                       // coset::CoseError does not implement std::Error::error
-                                       // and can't because coset is nostd.
+        coset_error: coset::CoseError, /* NOTE: We can not use #[transparent] here because
+                                        * coset::CoseError does not implement std::Error::error
+                                        * and can't because coset is nostd. */
```
Comment (chriskushmonotype) [inline sdk/src/error.rs:95]: (I presume this is just yet more fallout from the `rustfmt` settings.)

File: sdk/src/jumbf/boxes.rs [hunk 11]
```diff
@@ -2223,7 +2223,10 @@ impl BoxReader {
             }
         }
 
-        debug!("{}", format!("END#Label: {box_label:?}" /*jdesc.label()*/));
+        debug!(
+            "{}",
+            format!("END#Label: {box_label:?}" /* jdesc.label() */)
+        );
```
Comment (chriskushmonotype) [inline sdk/src/jumbf/boxes.rs:2229]: I find this needlessly fussy, but again I assume it's the `rustfmt`? (I note pedantically that `                debug!("{}", format!("END#Label: {box_label:?}" /* jdesc.label() */));` consumes only 78 columns.)

File: sdk/src/manifest_assertion.rs [hunk 12]
```diff
@@ -143,8 +143,8 @@ impl ManifestAssertion {
     ///```
     /// # use c2pa::Result;
     /// use c2pa::{
-    ///     assertions::{Actions, Action, c2pa_action},
-    ///     ManifestAssertion
+    ///     assertions::{c2pa_action, Action, Actions},
+    ///     ManifestAssertion,
```
Comment (chriskushmonotype) [inline sdk/src/manifest_assertion.rs:147]: ~Stray comma.~ "Oh right, Rust permits this..."

File: sdk/src/manifest_assertion.rs [hunk 13]
```diff
@@ -165,16 +165,16 @@ impl ManifestAssertion {
...
     /// };
     /// # fn main() -> Result<()> {
     /// let actions = Actions::new().add_action(Action::new(c2pa_action::EDITED));
     /// let manifest_assertion = ManifestAssertion::from_assertion(&actions)?;
     ///
     /// let actions: Actions = manifest_assertion.to_assertion()?;
     /// for action in actions.actions {
-    ///    println!("{}", action.action());
+    ///     println!("{}", action.action());
```
Comment (chriskushmonotype) [inline sdk/src/manifest_assertion.rs:177]: 🦅

File: sdk/tests/fixtures/font.otf
Comment (chriskushmonotype) [inline sdk/tests/fixtures/font.otf:1]: CFF huh.

File: sdk/src/asset_handlers/otf_io.rs [hunk 14]
```diff
@@ -183,118 +545,16 @@ impl AssetIO for OtfIO {
...
-        name_table.records.push(c2pa_name);
-        // Re-insert the modified name table.
-        font_file.tables.insert(name_table);
-        // Save back to the original file.
-        font_file
-            .save(asset_path)
-            .map_err(|_err| Error::FontSaveError)?;
-
-        Ok(())
+        add_c2pa_to_font(asset_path, store_bytes)
```
Comment (chriskushmonotype) [inline sdk/src/asset_handlers/otf_io.rs:548]: Or, _"A Farewell To CowPtrs"_.

**Review by chriskushmonotype — APPROVED** [review]
Exercised this by using the corresponding `c2patool` [fork PR #4](https://github.com/Monotype/c2patool/pull/4/files), and works as described, on TTFs and OTFs, with and without pre-existing C2PATOOL tables.

Interestingly, round-tripping `C2PA`-bearing files through `ttx` dis/re-assembly produces a font that no longer C2PA-verifies; probably because `ttx` can't un/re-wind the hashing (like it does for DSIG), right?

Behold the future!
image

File: sdk/src/error.rs (hunk 10 above)
Comment (brogdonm) [reply sdk/src/error.rs:95]: Yeah, this kind of stuff came from the recent upstream `main`.

Comment (brogdonm) [reply sdk/src/error.rs:95]: After discussions, going to close this PR and separate out the changes for a cleaner history.

File: sdk/tests/fixtures/font.otf
Comment (brogdonm) [reply sdk/tests/fixtures/font.otf:1]: haha, it was a fake font created from FontLab.

**Comment by brogdonm** [comment]
After discussions, going to close this PR and separate out the changes for a cleaner history.

**Pull request closed without merging**